vowpalwabbit/core/include/vw/core/global_data.h

```diff
@@ -239,77 +137,218 @@ class workspace
+  float power_t;  // the power on learning rate decay.
+  float eta;      // learning rate control.
+  float eta_decay_rate;
+};
+
+class loss_config
+{
+public:
+  std::unique_ptr<loss_function> loss;
+  float l1_lambda;  // the level of l_1 regularization to impose.
+  float l2_lambda;  // the level of l_2 regularization to impose.
+  bool no_bias;     // no bias in regularization
+  int reg_mode;
+};
+
+class reduction_state
+{
+public:
+  bool active;
+  bool bfgs;
   uint32_t lda;
+  // hack to support cb model loading into ccb learner
+  bool is_ccb_input_model = false;
+  void* /*Search::search*/ searchstr;
+  bool invariant_updates;  // Should we use importance aware/safe updates, gd only
+  uint32_t wpp;
+};
 
-  std::string text_regressor_name;
-  std::string inv_hash_regressor_name;
-  std::string json_weights_file_name;
-  bool dump_json_weights_include_feature_names = false;
-  bool dump_json_weights_include_extra_online_state = false;
+class runtime_config
+{
+public:
+  bool daemon;
+  bool vw_is_main = false;  // true if vw is executable; false in library mode
+  bool training;            // Should I train if lable data is available?
+  size_t pass_length;
+  size_t numpasses;
+  bool default_bits;
+  all_reduce_type selected_all_reduce_type;
+  uint32_t hash_seed;
+};
+
+class runtime_state
+{
+public:
+  VW::version_struct model_file_ver;
+  size_t passes_complete;
+  // Default value of 2 follows behavior of 1-indexing and can change to 0-indexing if detected
+  uint32_t indexing = 2;  // for 0 or 1 indexing
+  // bool nonormalize; not used?
+  bool do_reset_source;
+  std::unique_ptr<all_reduce_base> all_reduce;
+  VW::details::generate_interactions_object_cache generate_interactions_object_cache_state;
+  uint64_t parse_mask;  // 1 << num_bits -1
+};
+
+class parser_runtime
+{
+public:
+  std::string data_filename;
+  std::unique_ptr<parser> example_parser;
+  // Experimental field.
+  // Generic parser interface to make it possible to use any external parser.
+  std::unique_ptr<VW::details::input_parser> custom_parser;
+  std::thread parse_thread;
+  size_t max_examples;  // for TLC
+  bool chain_hash_json = false;
+#ifdef BUILD_FLATBUFFERS
+  std::unique_ptr<VW::parsers::flatbuffer::parser> flat_converter;
+#endif
+};
+
+class output_config
+{
+public:
+  bool quiet;
+  bool audit;  // should I print lots of debugging information?
+  bool hash_inv;
+  bool print_invert;
+  bool hexfloat_weights;
+};
+
+class output_runtime
+{
+public:
+  // error reporting
+  std::shared_ptr<details::trace_message_wrapper> trace_message_wrapper_context;
+  std::shared_ptr<std::ostream> trace_message;
 
-  size_t length() { return (static_cast<size_t>(1)) << num_bits; };
+  std::unique_ptr<VW::io::writer> stdout_adapter;
+
+  std::map<uint64_t, VW::details::invert_hash_info> index_name_map;
+  std::shared_ptr<std::vector<char>> audit_buffer;
+  std::unique_ptr<VW::io::writer> audit_writer;
+  VW::metrics_collector global_metrics;

```

Comment:
same question for some of the loggers and metrics, could someone running in lib mode want to consume?